### PR TITLE
Cgroup coverage for RedHat and Centos

### PR DESCRIPTION
**Pull request by nagworld9**
## Description
Cgroup coverage for RedHat and CentOs

---

### PR information
- [ ] The title of the PR is clear and informative.
- [ ] There are a small number of commits, each of which has an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).
- [ ] If applicable, the PR references the bug/issue that it fixes in the description.
- [ ] New Unit tests were added for the changes made

### Quality of Code and Contribution Guidelines
- [ ] I have read the [contribution guidelines](https://github.com/Azure/WALinuxAgent/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: azurelinuxagent/common/cgroupapi.py [hunk 1]
```diff
@@ -54,9 +57,12 @@ def cgroups_supported():
         distro_name = distro_info[0]
         try:
             distro_version = FlexibleVersion(distro_info[1])
+            major_minor_version = "{0}.{1}".format(distro_version.major, distro_version.minor)
         except ValueError:
             return False
-        return distro_name.lower() == 'ubuntu' and distro_version.major >= 16
+        return ((distro_name.lower() == 'ubuntu' and distro_version.major >= 16) or
+                (distro_name.lower() in ("centos", "redhat") and distro_version.major >= 7 and
+                 major_minor_version != "7.1"))
```
Comment (nagworld9) [inline azurelinuxagent/common/cgroupapi.py:65]: we stated (in setup.py) there is a systemctl bug on 7.1 and also, I don't find 7.1 image form marketplace for testing. So I marked 7.1 not to enable.

File: setup.py [hunk 2]
```diff
@@ -167,8 +177,8 @@ def get_data_files(name, version, fullname):  # pylint: disable=R0912
             set_systemd_files(data_files, dest=systemd_dir_path,
                               src=[
                                   "init/ubuntu/walinuxagent.service",
-                                  "init/ubuntu/azure.slice",
-                                  "init/ubuntu/azure-vmextensions.slice"
+                                  "init/azure.slice",
+                                  "init/azure-vmextensions.slice"
                               ])
```
Comment (nagworld9) [inline setup.py:182]: Slices are common for all distros so moved to common folder

File: setup.py [hunk 3]
```diff
@@ -108,12 +109,21 @@ def get_data_files(name, version, fullname):  # pylint: disable=R0912
         if version.startswith("8.2"):
             # redhat 8+ uses systemd and python3
             set_systemd_files(data_files, dest=systemd_dir_path,
-                              src=["init/redhat/waagent.service"])
+                              src=["init/redhat/py3/waagent.service",
```
Comment (nagworld9) [inline setup.py:112]: created py3 folder to keep py3 dependency service.

File: azurelinuxagent/common/cgroupapi.py (hunk 1 above)
Comment (narrieta) [reply azurelinuxagent/common/cgroupapi.py:65]: what is the issue with 7.1?

File: azurelinuxagent/common/cgroupapi.py [hunk 4]
```diff
@@ -253,7 +260,8 @@ def get_extension_slice_name(extension_name):
         # Since '-' is used as a separator in systemd unit names, we replace it with '_' to prevent side-effects.
         return EXTENSION_SLICE_PREFIX + "-" + extension_name.replace('-', '_') + ".slice"
 
-    def start_extension_command(self, extension_name, command, cmd_name, timeout, shell, cwd, env, stdout, stderr, error_code=ExtensionErrorCodes.PluginUnknownFailure):
+    def start_extension_command(self, extension_name, command, cmd_name, timeout, shell, cwd, env, stdout, stderr,
```
Comment (narrieta) [inline azurelinuxagent/common/cgroupapi.py:263]: next time, could you separate formatting-only changes from code changes? (unless there's just a couple of them) thanks

File: init/redhat/py3/waagent.service [hunk 5]
```diff
@@ -0,0 +1,18 @@
+[Unit]
```
Comment (narrieta) [inline init/redhat/py3/waagent.service:1]: how did you get the contents of this file?

File: init/redhat/waagent.service [hunk 6]
```diff
@@ -8,9 +8,11 @@ ConditionPathExists=/etc/waagent.conf
 
 [Service]
 Type=simple
-ExecStart=/usr/bin/python3 -u /usr/sbin/waagent -daemon
+ExecStart=/usr/bin/python -u /usr/sbin/waagent -daemon
```
Comment (narrieta) [inline init/redhat/waagent.service:11]: why this change?

File: init/redhat/py3/waagent.service (hunk 5 above)
Comment (nagworld9) [reply init/redhat/py3/waagent.service:1]: Context: We already have this file under `redhat` folder to handle py3 distro and also, we use `init `folder service file for py2. As part of this change, I have to modify these unit files to add slice info. Since `init `folder unit file being shared between all other distros, I moved original py3 file from `redhat `to `redhat/py3` path and copied common service to `redhat `folder. Hope this answered your question.

File: init/redhat/waagent.service (hunk 6 above)
Comment (nagworld9) [reply init/redhat/waagent.service:11]: I explained in other comment.

File: init/redhat/py3/waagent.service (hunk 5 above)
Comment (narrieta) [reply init/redhat/py3/waagent.service:1]: yup, thanks

File: azurelinuxagent/common/cgroupapi.py (hunk 1 above)
Comment (nagworld9) [reply azurelinuxagent/common/cgroupapi.py:65]: I don't know the issue. I'm going by how setup.py was implemented. This was the comment stated.
```
            if version.startswith("7.1"):
                # TODO this is a mitigation to systemctl bug on 7.1
                set_sysv_files(data_files)
```

File: init/redhat/waagent.service (hunk 6 above)
Comment (narrieta) [reply init/redhat/waagent.service:11]: since this the "main" file, I'd leave the Python version as 3 here, and create a py2 directory for the older distros (instead of py3)

File: azurelinuxagent/common/cgroupapi.py (hunk 1 above)
Comment (narrieta) [reply azurelinuxagent/common/cgroupapi.py:65]: can you investigate? what happens if you just enable it?

Comment (nagworld9) [reply azurelinuxagent/common/cgroupapi.py:65]: As I said, I don't find image from marketplace to test it.

Comment (nagworld9) [reply azurelinuxagent/common/cgroupapi.py:65]: Maybe I can try if DCR can create the VM

Comment (narrieta) [reply azurelinuxagent/common/cgroupapi.py:65]: what are the minimum versions tested in DCR and endorsed? maybe we should limit to those

Comment (nagworld9) [reply azurelinuxagent/common/cgroupapi.py:65]: Minimum version tested `7.8`

Comment (narrieta) [reply azurelinuxagent/common/cgroupapi.py:65]: how about endorsed?

Comment (nagworld9) [reply azurelinuxagent/common/cgroupapi.py:65]: https://docs.microsoft.com/en-us/azure/virtual-machines/linux/endorsed-distros

supported versions `6.x, 7.x, 8.x`

Comment (narrieta) [reply azurelinuxagent/common/cgroupapi.py:65]: ok, then let's add to DCR the minimum endorsed version that we can test and then enable the feature only on that and later versions (i.e. we should not enable this feature unless it is tested)

Comment (kevinclark19a) [inline azurelinuxagent/common/cgroupapi.py:65]: this seems like it could get unwieldy... Can we move this into a function inside the OSUtil class heirarchy? that way we can keep this logic close to the distro itself.

File: azurelinuxagent/common/cgroupconfigurator.py [hunk 7]
```diff
@@ -755,12 +768,6 @@ def remove_extension_slice(self, extension_name):
                 if os.path.exists(extension_slice_path):
                     self.stop_tracking_extension_cgroups(extension_name)
                     CGroupConfigurator._Impl.__cleanup_unit_file(extension_slice_path)
-                # stop the unit gracefully; the extensions slices will be removed from /sys/fs/cgroup path
-                try:
-                    logger.info("Executing systemctl stop {0}".format(extension_slice_name))
-                    shellutil.run_command(["systemctl", "stop", extension_slice_name])
-                except Exception as exception:
-                    _log_cgroup_warning("systemctl stop failed (remove slice): {0}", ustr(exception))
```
Comment (kevinclark19a) [inline azurelinuxagent/common/cgroupconfigurator.py:763]: It's not clear to me why this code was removed... What is the reasoning here?

File: azurelinuxagent/common/cgroupapi.py (hunk 1 above)
Comment (nagworld9) [reply azurelinuxagent/common/cgroupapi.py:65]: @kevinclark19a I like the idea, moving to each OSUtil and create abstract function quite bit of change, I will keep in mind next time.

File: azurelinuxagent/common/cgroupconfigurator.py (hunk 7 above)
Comment (nagworld9) [reply azurelinuxagent/common/cgroupconfigurator.py:763]: Regularly, it's the system that cleans up these slices if the croups are not active. 

This run is to remove the slices from /sys/fs/cgroup explicitly if still present. This becomes flaky and not working all the time and also, redhat and centos complaining about access denied while running this command. So, I decided to remove and relay on OS to preform cleanup.

File: init/redhat/waagent.service (hunk 6 above)
Comment (nagworld9) [reply init/redhat/waagent.service]: updated

File: azurelinuxagent/common/cgroupapi.py (hunk 1 above)
Comment (nagworld9) [reply azurelinuxagent/common/cgroupapi.py]: updated

**Review by kevinclark19a — APPROVED** [review]

File: azurelinuxagent/common/cgroupconfigurator.py [hunk 8]
```diff
@@ -510,6 +518,7 @@ def __reset_agent_cpu_quota():
             logger.info("Resetting agent's CPUQuota")
             if CGroupConfigurator._Impl.__try_set_cpu_quota(''):  # setting an empty value resets to the default (infinity)
                 CGroupsTelemetry.set_track_throttled_time(False)
+                _log_cgroup_info('CPUQuota: {0}', systemd.get_unit_property(systemd.get_agent_unit_name(), "CPUQuotaPerSecUSec"))
```
Comment (narrieta) [inline azurelinuxagent/common/cgroupconfigurator.py:521]: aren't we logging this at line 426?

Comment (nagworld9) [reply azurelinuxagent/common/cgroupconfigurator.py:521]: yes, that will be logged when cgroup initializes but not after resetting the quota.

Comment (narrieta) [reply azurelinuxagent/common/cgroupconfigurator.py:521]: ok

**Review by narrieta — APPROVED** [review]
